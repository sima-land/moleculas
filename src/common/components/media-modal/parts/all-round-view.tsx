import { useCallback, useState } from 'react';
import { useBreakpoint } from '@sima-land/ui-nucleons/hooks/breakpoint';
import {
  AllRoundView as BaseAllRoundView,
  AllRoundViewProps,
  ViewState,
} from '../../../../desktop/components/gallery-modal/components/all-round-view';
import { useMounted } from '../../../../common/hooks';
import TouchSVG from '@sima-land/ui-quarks/icons/64x64/Colorful/TouchScrollX';
import styles from './all-round-view.module.scss';

/**
 * Просмотр 360.
 * @param props Свойства.
 * @return Элемент.
 */
export function AllRoundView(props: AllRoundViewProps) {
  const mounted = useMounted();
  const desktop = useBreakpoint('xs+');

  if (!mounted) {
    return null;
  }

  return <AfterMount desktop={desktop} {...props} />;
}

/**
 * Внутренний компонент который будет выведен только после монтирования.
 * @param props Свойства.
 * @return Элемент.
 */
function AfterMount({ desktop, ...props }: AllRoundViewProps & { desktop: boolean }) {
  const [state, setState] = useState<ViewState>(() => (desktop ? 'autoplay' : 'default'));
  const [tutorial, setTutorial] = useState(() => !desktop);

  const acceptTutorial = useCallback(() => {
    setTutorial(false);
    setState('autoplay');
  }, []);

  return (
    <div className={styles.root}>
      {!desktop && tutorial && (
        <div className={styles.tutorial} onPointerUp={acceptTutorial}>
          <TouchSVG />
        </div>
      )}

      <BaseAllRoundView {...props} controls={desktop} state={state} onStateChange={setState} />
    </div>
  );
}
