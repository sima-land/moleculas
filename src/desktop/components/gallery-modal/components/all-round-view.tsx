import { useState, useRef, useEffect } from 'react';
import { useIdentityRef } from '@sima-land/ui-nucleons/hooks';
import { on } from '@sima-land/ui-nucleons/helpers';
import { useImagesLoad } from '../utils';
import classNames from 'classnames/bind';
import styles from './all-round-view.m.scss';
import AllRoundSVG from '@sima-land/ui-quarks/icons/40x40/Filled/Viewing360Perspective';
import PauseSVG from '@sima-land/ui-quarks/icons/40x40/Filled/Pause';
import TurnLeftSVG from '@sima-land/ui-quarks/icons/16x16/Stroked/ArrowTurnLeft';
import TurnRightSVG from '@sima-land/ui-quarks/icons/16x16/Stroked/ArrowTurnRight';
import { ImgStub } from '../../../../common/components/img-stub';
import { Hint, useHintFloating, useHintOnHover } from '@sima-land/ui-nucleons/hint';

export interface AllRoundViewProps {
  photos: string[];
  controls?: boolean;
  state?: ViewState;
  onStateChange?: (state: ViewState) => void;
}

export type ViewState = 'default' | 'autoplay' | 'turn-right' | 'turn-left';

const cx = classNames.bind(styles);

/**
 * 3д-просмотр.
 * @param props Свойства.
 * @return Элемент.
 */
export const AllRoundView = ({
  photos,
  controls = true,
  state: stateProp = 'autoplay',
  onStateChange,
}: AllRoundViewProps) => {
  const imageRef = useRef<HTMLImageElement>(null);
  const dragStartRef = useRef<{ position: number; index: number } | null>(null);

  const [state, setState] = useState<ViewState>(stateProp);

  useEffect(() => onStateChange?.(state), [state]);
  useEffect(() => setState(stateProp), [stateProp]);

  const autoplay = state === 'autoplay';

  // eslint-disable-next-line jsdoc/require-jsdoc
  const ChangeState = (bindState: ViewState) => () => setState(bindState);

  const [index, setIndex] = useState<number>(0);
  const indexRef = useIdentityRef(index);
  const imagesStatus = useImagesLoad(photos);
  const ready = imagesStatus === 'done';
  const failed = imagesStatus === 'fail';

  // запуск/остановка вращения
  useEffect(() => {
    if (state !== 'default' && ready) {
      const timerId = window.setInterval(() => {
        const increment = state === 'turn-left' ? -1 : 1;

        setIndex((photos.length + (indexRef.current + increment)) % photos.length);
      }, 1000 / 30);

      return () => window.clearInterval(timerId);
    }
  }, [state, ready]);

  // перетаскивание
  useEffect(() => {
    // переданное число всегда замениться на такое, которое будет в промежутке от 0 до количества фото
    // eslint-disable-next-line jsdoc/require-jsdoc
    const wrap = (value: number) => (photos.length + value) % photos.length;

    const image = imageRef.current;

    if (!image) {
      return;
    }

    const offList = [
      on<PointerEvent>(image, 'pointerdown', e => {
        e.preventDefault();
        e.stopPropagation();

        // ВАЖНО: preventDefault предотвращает расфокусировку полей ввода - форсируем
        if (document.activeElement instanceof HTMLElement) {
          document.activeElement.blur();
        }

        setState('default');

        const imageLeft = image.getBoundingClientRect().left;
        image.style.touchAction = 'none';

        dragStartRef.current = {
          position: e.clientX - imageLeft,
          index: indexRef.current,
        };
      }),

      on<PointerEvent>(window, 'pointermove', e => {
        if (dragStartRef.current !== null) {
          const imageWidth = image.clientWidth;
          const imageLeft = image.getBoundingClientRect().left;
          const startX = dragStartRef.current.position;
          const currentX = e.clientX - imageLeft;

          const startPos = wrap(Math.round((startX * photos.length) / imageWidth));
          const currentPos = wrap(Math.round((currentX * photos.length) / imageWidth));
          const nextIndex = wrap(dragStartRef.current.index + (currentPos - startPos));

          setIndex(nextIndex);
        }
      }),

      on<PointerEvent>(window, 'pointerup', () => {
        dragStartRef.current = null;
        image.style.touchAction = '';
      }),
    ];

    return () => offList.forEach(fn => fn());
  }, [photos.length, failed]);

  // состояние
  const [open, setOpen] = useState<boolean>(false);

  // позиционирование
  const { refs, ...floating } = useHintFloating({ open, onOpenChange: setOpen });

  // пользовательское взаимодействие
  const { getReferenceProps, getFloatingProps } = useHintOnHover(floating);

  return (
    <div className={cx('root')}>
      {failed && <ImgStub className={cx('stub')} />}
      {!failed && (
        <img data-testid='gallery-modal:360-current-photo' ref={imageRef} src={photos[index]} />
      )}

      {!failed && controls && (
        <div className={cx('controls')}>
          <button
            className={cx('control', 'turn')}
            disabled={autoplay}
            onPointerDown={ChangeState('turn-left')}
            onPointerUp={ChangeState('default')}
            onPointerLeave={ChangeState('default')}
            data-testid='gallery-modal:360-turn-left-button'
          >
            <TurnRightSVG />
          </button>

          <button
            ref={refs.setReference}
            {...getReferenceProps()}
            className={cx('control', 'primary', autoplay && 'pause')}
            onClick={() => {
              setState(s => (s === 'autoplay' ? 'default' : 'autoplay'));
            }}
            data-testid='gallery-modal:toggle-autoplay-button'
          >
            {autoplay ? <PauseSVG /> : <AllRoundSVG />}
          </button>

          {autoplay && (
            <Hint
              open={open}
              hintRef={refs.setFloating}
              arrowRef={refs.setArrow}
              {...getFloatingProps()}
            >
              <>
                Вы можете вращать товар,
                <br />
                зажав левую кнопку
                <br />
                мыши на фотографии
              </>
            </Hint>
          )}

          <button
            className={cx('control', 'turn')}
            disabled={autoplay}
            onPointerDown={ChangeState('turn-right')}
            onPointerUp={ChangeState('default')}
            onPointerLeave={ChangeState('default')}
            data-testid='gallery-modal:360-turn-right-button'
          >
            <TurnLeftSVG />
          </button>
        </div>
      )}
    </div>
  );
};
