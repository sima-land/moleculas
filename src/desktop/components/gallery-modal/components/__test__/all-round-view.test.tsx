import React from 'react';
import { mount } from 'enzyme';
import { AllRoundView } from '../all-round-view';
import { act } from 'react-dom/test-utils';

jest.mock('../../utils', () => {
  const original = jest.requireActual('../../utils');

  return {
    ...original,
    __esModule: true,
    useImagesLoad: jest.fn(() => true),
  };
});

describe('AllRoundView', () => {
  const testPhotos = [
    'https://img.com/0',
    'https://img.com/1',
    'https://img.com/2',
    'https://img.com/3',
    'https://img.com/4',
    'https://img.com/5',
    'https://img.com/6',
    'https://img.com/7',
    'https://img.com/8',
    'https://img.com/9',
  ];

  const Selectors = {
    image: '[data-testid="gallery-modal:360-current-photo"]',
    autoplayButton: '[data-testid="gallery-modal:toggle-autoplay-button"]',
    turnLeftButton: '[data-testid="gallery-modal:360-turn-left-button"]',
    turnRightButton: '[data-testid="gallery-modal:360-turn-right-button"]',
    hint: '[data-testid="hint"]',
  } as const;

  it('should renders correctly', () => {
    const wrapper = mount(<AllRoundView photos={[]} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should renders correctly default state', () => {
    const wrapper = mount(<AllRoundView photos={testPhotos} />);

    act(() => {
      wrapper.find(Selectors.autoplayButton).simulate('click');
    });
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
  });

  it('should change image src every 1000 / 30 ms', () => {
    jest.useFakeTimers();

    const wrapper = mount(<AllRoundView photos={testPhotos} />);

    expect(wrapper.find(Selectors.image).prop('src')).toEqual(testPhotos[0]);

    act(() => {
      jest.advanceTimersByTime(1000 / 30 + 10);
    });
    wrapper.update();

    expect(wrapper.find(Selectors.image).prop('src')).toEqual(testPhotos[1]);

    act(() => {
      jest.advanceTimersByTime(1000 / 30 + 10);
    });
    wrapper.update();

    expect(wrapper.find(Selectors.image).prop('src')).toEqual(testPhotos[2]);

    // stop autoplay
    act(() => {
      wrapper.find(Selectors.autoplayButton).simulate('click');
    });
    wrapper.update();

    act(() => {
      jest.advanceTimersByTime(1000 / 30 + 10);
    });
    wrapper.update();

    expect(wrapper.find(Selectors.image).prop('src')).toEqual(testPhotos[2]);
  });

  it('should handle pointer down on turn left button', () => {
    jest.useFakeTimers();

    const wrapper = mount(<AllRoundView photos={testPhotos} />);

    act(() => {
      wrapper.find(Selectors.turnLeftButton).simulate('pointerdown');
    });
    wrapper.update();

    expect(wrapper.find(Selectors.image).prop('src')).toEqual(testPhotos[0]);

    act(() => {
      jest.advanceTimersByTime(1000 / 30 + 10);
    });
    wrapper.update();

    expect(wrapper.find(Selectors.image).prop('src')).toEqual(testPhotos[9]);

    act(() => {
      jest.advanceTimersByTime(1000 / 30 + 10);
    });
    wrapper.update();

    expect(wrapper.find(Selectors.image).prop('src')).toEqual(testPhotos[8]);
  });

  it('should handle pointer down on turn right button', () => {
    jest.useFakeTimers();

    const wrapper = mount(<AllRoundView photos={testPhotos} />);

    act(() => {
      wrapper.find(Selectors.turnRightButton).simulate('pointerdown');
    });
    wrapper.update();

    expect(wrapper.find(Selectors.image).prop('src')).toEqual(testPhotos[0]);

    act(() => {
      jest.advanceTimersByTime(1000 / 30 + 10);
    });
    wrapper.update();

    expect(wrapper.find(Selectors.image).prop('src')).toEqual(testPhotos[1]);

    act(() => {
      jest.advanceTimersByTime(1000 / 30 + 10);
    });
    wrapper.update();

    expect(wrapper.find(Selectors.image).prop('src')).toEqual(testPhotos[2]);
  });

  it('should handle autoplay button click', () => {
    jest.useFakeTimers();

    const wrapper = mount(<AllRoundView photos={testPhotos} />);

    act(() => {
      wrapper.find(Selectors.turnRightButton).simulate('pointerdown');
    });
    wrapper.update();

    expect(wrapper.find(Selectors.image).prop('src')).toEqual(testPhotos[0]);

    act(() => {
      jest.advanceTimersByTime(1000 / 30 + 10);
    });
    wrapper.update();

    expect(wrapper.find(Selectors.image).prop('src')).toEqual(testPhotos[1]);

    act(() => {
      jest.advanceTimersByTime(1000 / 30 + 10);
    });
    wrapper.update();

    expect(wrapper.find(Selectors.image).prop('src')).toEqual(testPhotos[2]);

    // autoplay button click
    act(() => {
      wrapper.find(Selectors.autoplayButton).simulate('click');
    });
    wrapper.update();

    act(() => {
      jest.advanceTimersByTime(1000 / 30 + 10);
    });
    wrapper.update();

    expect(wrapper.find(Selectors.image).prop('src')).toEqual(testPhotos[3]);
  });

  it('should show/hide hint', () => {
    const wrapper = mount(<AllRoundView photos={testPhotos} />);

    expect(wrapper.find(Selectors.hint)).toHaveLength(0);

    act(() => {
      wrapper.find(Selectors.autoplayButton).simulate('mouseenter');
    });
    wrapper.update();

    expect(wrapper.find(Selectors.hint)).toHaveLength(1);

    act(() => {
      wrapper.find(Selectors.autoplayButton).simulate('mouseleave');
    });
    wrapper.update();

    expect(wrapper.find(Selectors.hint)).toHaveLength(0);
  });

  it('should handle image pointerdown, window pointerup', () => {
    const wrapper = mount(<AllRoundView photos={testPhotos} />);

    act(() => {
      wrapper
        .find(Selectors.image)
        .getDOMNode()
        .dispatchEvent(new Event('pointerdown', { clientX: 10 } as any));
    });
    wrapper.update();

    expect(wrapper.find(Selectors.image).getDOMNode<HTMLElement>().style.touchAction).toBe('none');

    act(() => {
      window.dispatchEvent(new Event('pointerup'));
    });
    wrapper.update();

    expect(wrapper.find(Selectors.image).getDOMNode<HTMLElement>().style.touchAction).toBe('');
  });

  it('should handle window pointerup', () => {
    jest.useFakeTimers();

    jest.spyOn(HTMLImageElement.prototype, 'clientWidth', 'get').mockImplementation(() => 100);
    jest.spyOn(HTMLImageElement.prototype, 'getBoundingClientRect').mockImplementation(
      () =>
        ({
          left: 0,
        } as any),
    );

    const wrapper = mount(<AllRoundView photos={testPhotos} />);

    expect(wrapper.find(Selectors.image).prop('src')).toEqual(testPhotos[0]);

    // move without pointer down
    act(() => {
      window.dispatchEvent(new MouseEvent('pointermove', { clientX: 20 } as any));
    });
    wrapper.update();

    expect(wrapper.find(Selectors.image).prop('src')).toEqual(testPhotos[0]);

    // pointer down
    act(() => {
      wrapper
        .find(Selectors.image)
        .getDOMNode()
        .dispatchEvent(new MouseEvent('pointerdown', { clientX: 2 }));
    });
    wrapper.update();

    // move
    act(() => {
      window.dispatchEvent(new MouseEvent('pointermove', { clientX: 93 } as any));
    });
    wrapper.update();

    expect(wrapper.find(Selectors.image).prop('src')).toEqual(testPhotos[9]);

    wrapper.unmount();
  });
});
