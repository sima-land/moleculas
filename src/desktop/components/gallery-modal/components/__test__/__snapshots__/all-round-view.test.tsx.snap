// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllRoundView should renders correctly default state 1`] = `
<div>
  <div
    class="root"
  >
    <img
      data-testid="gallery-modal:360-current-photo"
      src="https://img.com/0"
    />
    <div
      class="controls"
    >
      <button
        class="control turn"
        data-testid="gallery-modal:360-turn-left-button"
      >
        <svg
          height="16"
          viewBox="0 0 16 16"
          width="16"
        >
          <g
            clip-path="url(#prefix__clip0_5876_1440)"
          >
            <path
              clip-rule="evenodd"
              d="M4 1a1 1 0 0 1 1-1h6a1 1 0 0 1 .928 1.371l-2 5a1 1 0 0 1-1.856-.742L9.22 2.757C7.257 3.88 6 5.839 6 8c0 2.829 2.153 5.311 5.227 6.026a1 1 0 0 1-.453 1.948C6.946 15.084 4 11.907 4 8c0-2.393 1.106-4.512 2.842-6H5a1 1 0 0 1-1-1"
              fill-rule="evenodd"
            />
          </g>
          <defs>
            <clippath
              id="prefix__clip0_5876_1440"
            >
              <path
                d="M0 0h16v16H0z"
              />
            </clippath>
          </defs>
        </svg>
      </button>
      <button
        class="control primary"
        data-testid="gallery-modal:toggle-autoplay-button"
      >
        <svg
          height="40"
          viewBox="0 0 40 40"
          width="40"
        >
          <path
            clip-rule="evenodd"
            d="M30.63 17.879q.861-1.258.861-3.876 0-2.53-.889-3.842-.881-1.312-2.61-1.312-1.779 0-2.633 1.257-.855 1.251-.854 3.897 0 2.535.875 3.835.882 1.299 2.611 1.299 1.785 0 2.639-1.258m-3.712-1.237q-.315-.8-.314-2.64 0-1.837.32-2.638.322-.807 1.067-.806.732 0 1.06.806.335.8.335 2.639 0 1.818-.328 2.625-.33.807-1.067.807-.752 0-1.073-.793m-11.758-5.4q0 .935-.567 1.592-.569.657-1.593.902v.041q1.21.15 1.832.739.622.58.622 1.572 0 1.442-1.046 2.249-1.046.8-2.987.8-1.627 0-2.885-.54v-1.798q.582.294 1.278.478.698.185 1.381.185 1.047 0 1.545-.356.5-.355.5-1.141 0-.704-.575-.998-.574-.294-1.832-.294h-.759v-1.62h.773q1.162 0 1.695-.301.54-.308.54-1.046 0-1.135-1.422-1.135-.492 0-1.005.164-.505.165-1.128.568L8.55 9.847q1.367-.985 3.26-.985 1.552 0 2.448.63.902.627.902 1.75m1.333 3.513q0-2.967 1.251-4.416 1.258-1.45 3.76-1.45.854 0 1.34.103v1.689a5.5 5.5 0 0 0-1.203-.137q-1.087 0-1.778.328a2.2 2.2 0 0 0-1.025.97q-.342.644-.403 1.826h.088q.678-1.162 2.167-1.162 1.34 0 2.1.84.758.842.758 2.325 0 1.6-.902 2.536-.903.93-2.502.93-1.108 0-1.935-.513-.82-.512-1.271-1.497-.445-.984-.445-2.372m3.61 2.693q.675 0 1.039-.45.362-.46.362-1.3 0-.73-.342-1.148-.335-.424-1.018-.424-.643 0-1.101.417-.45.418-.451.97 0 .814.424 1.375.43.56 1.087.56m14.923-3.141a1 1 0 1 0-1.052 1.701c1.155.713 1.902 1.45 2.362 2.2.456.74.664 1.543.664 2.45 0 1.704-1.251 3.48-3.813 4.979-2.526 1.479-6.123 2.552-10.27 2.9a1 1 0 0 0 .167 1.992c4.365-.366 8.268-1.5 11.114-3.166 2.81-1.645 4.802-3.95 4.802-6.706 0-1.236-.291-2.41-.96-3.498-.664-1.079-1.667-2.02-3.014-2.852m-29 2.044a1 1 0 1 0-1.052-1.702c-1.31.81-2.316 1.652-2.99 2.644C1.296 18.308 1 19.413 1 20.657c0 2.755 1.992 5.06 4.802 6.706 2.332 1.364 5.373 2.373 8.797 2.894l-1.746 1.48a1 1 0 0 0 1.294 1.526l3.5-2.967a1 1 0 0 0 .115-1.41l-3-3.533a1 1 0 0 0-1.524 1.294l1.343 1.583c-3.09-.503-5.765-1.42-7.768-2.593C4.25 24.137 3 22.362 3 20.657c0-.899.204-1.6.64-2.24.45-.666 1.195-1.33 2.386-2.066"
            fill-rule="evenodd"
          />
        </svg>
      </button>
      <button
        class="control turn"
        data-testid="gallery-modal:360-turn-right-button"
      >
        <svg
          height="16"
          viewBox="0 0 16 16"
          width="16"
        >
          <g
            clip-path="url(#prefix__clip0_5876_1441)"
          >
            <path
              clip-rule="evenodd"
              d="M12 1a1 1 0 0 0-1-1H5a1 1 0 0 0-.928 1.371l2 5a1 1 0 0 0 1.856-.742L6.78 2.757C8.743 3.88 10 5.839 10 8c0 2.829-2.153 5.311-5.226 6.026a1 1 0 0 0 .452 1.948C9.055 15.084 12 11.907 12 8c0-2.393-1.106-4.512-2.842-6H11a1 1 0 0 0 1-1"
              fill-rule="evenodd"
            />
          </g>
          <defs>
            <clippath
              id="prefix__clip0_5876_1441"
            >
              <path
                d="M0 0h16v16H0z"
              />
            </clippath>
          </defs>
        </svg>
      </button>
    </div>
  </div>
</div>
`;
